### DISCUSSION TITLE
What triggers recompilation for sparse input?

### DISCUSSION THREAD

**Discussion by krzysztofrusek** [jax-ml/jax#19658] [Q&A] [answered]
In  simple [trials](https://colab.research.google.com/gist/krzysztofrusek/3eeea74dd6acfd0b7f48acecb9d8670b/scratchpad.ipynb) I have found that recompilation is triggered based on the shape of array i.e. just like for dense counterparts.
I just want to make sure if this is correct? 

My second question is what is the recommended way to grow sparse array to a given shape ( not reshaping) as `jnp.pad` is not supported?

**Comment by jakevdp** [comment, marked as answer]
If you're talking about `jax.experimental.sparse` objects: compilation is cached based on the static metadata (i.e. shape and dtype) as well as changes to the shape and dtype of the dense buffers representing the sparse matrix.

Regarding your second question: it would be helpful if you could show example code demonstrating the type of operation you'd like to do.

**Reply by krzysztofrusek** [reply]
Thank you, the first part is solved.

Regarding the second question, I want to use this in GNN implementation so I wanted to add extra nodes and edges to the graph just like  `jraph` does to avoid recompilation for new graphs.

The operation can be described as

```python
B = jnp.zeros((6,6)).at[2,1].set(3).at[4,3].set(1).at[4,5].set(1)
sB = sparse.BCOO.fromdense(B)

bigB=sparse.BCOO((jnp.pad(sB.data,...),jnp.pad(sB.indices,....)), shape=(16,16))
```

Also construction a large sparse block diagonal from sparse matrixes would be useful for graph batching.

**Reply by jakevdp** [reply]
If you add new nodes to the graph, you're going to trigger recompilation – there's no way around that.

**Reply by jakevdp** [reply]
Actually, I realized there is a way around that: you could pre-allocate the full graph, but mask out the part of the matrix you don't need (any index value greater or equal to the axis dimension will be treated as a masked value by the sparse implementations) then when you add new nodes, you could add them by converting these masked entries into defined entries.

There are no high-level utilities for this, so to accomplish it you'd have to work with the underlying `data` and `indices` buffers directly. Hope that helps!

**Reply by krzysztofrusek** [reply]
Interesting idea, and not that obvious. Maybe you should add this to the documentation of sparse module together with your previous answer about recompilation.

**Reply by jakevdp** [reply]
Yeah – the sparse docs are pretty incomplete, but it's an experimental implementation and we don't intend full support.

**Reply by krzysztofrusek** [reply]
Any plans to make it stable? Otherwise everything based on `jax.sparse` will alse be experimental. 
This equation started with the intention simplify the interface to my library [queuinx](https://github.com/krzysztofrusek/queuinx) and use it once sparse is stable.

**Reply by jakevdp** [reply]
At this point there are no definite plans for further development of `jax.experimental.sparse`. The main problem is that XLA has no sparse primitives, and so the performance of the sparse operations is very slow. Given that, we did not find enough real-world use-cases of the tools to justify further development.
